File: sdks/python/cognitum/meta_proxy/nonstream.py
"""Non-streaming POST forwarding for the Meta Proxy data plane (ADR-0025a §D7).

This is a lightweight, Proxy-SPECIFIC equivalent of
:mod:`cognitum.meta_llm.nonstream`'s ``post_json_idempotent`` rather than a
reuse of it: that helper is bound to ``MetaLlmClientConfig`` (its
``credential_provider``/``base_url``/``MetaLlmTelemetryHooks``), whereas the
Proxy has its own config (``local_credential_provider``/``origin``/
``MetaProxyTelemetryHooks``) AND three Proxy-only obligations that helper
does not carry:

1. the §D7 caller-header FORWARDING ALLOWLIST -- only an explicit set of
   headers may be forwarded, and ``Authorization``/local-bearer/``Host``/
   ``Content-Length``/sponsor/identity/consent are NEVER caller-forwarded
   (they come from validated local state);
2. §D10 redirect rejection -- a 3xx is a hard protocol error, never followed;
3. §D4/§D5 routing-receipt decode into ``MetaProxyResponseMeta`` so the
   caller's ``RoutingIntent`` can be verified at decode time.

The idempotency-key + single-401-refresh shape is mirrored CONCEPTUALLY from
``meta_llm/nonstream.py``, but the retry shape is deliberately NOT mirrored:
per §D8, "No Proxy POST is automatically retried while it drops
``Idempotency-Key``". That sentence is about whether the *Proxy server*
honors the header for server-side deduplication -- the currently-deployed
Proxy drops it -- not about whether the SDK attaches one. Attaching an
``Idempotency-Key`` client-side does nothing to make a retry safe if the
server never uses it to deduplicate, so a 429/502/503 (or any other
non-2xx) from this POST is surfaced as a single terminal, non-retryable
error; the caller may retry manually using the error's ``retry_after_ms``.
The Alternatives-considered table makes the same point explicitly: "Retry
Proxy POSTs" was considered and rejected because "Idempotency is dropped
and spend can duplicate". Bounded retry remains reserved for the read-only
status/models/identity routes (§D8), which this module does not implement.
§D8's own error/stream/deadline model is otherwise out of scope for this
pass.
"""

from __future__ import annotations

import time
import uuid
from collections.abc import Mapping
from typing import TYPE_CHECKING, Any

import httpx

from cognitum.agentic import AgenticError
from cognitum.meta_proxy.envelope import MetaProxyResponseMeta, collect_unknown_headers
from cognitum.meta_proxy.http_errors import map_meta_proxy_http_error
from cognitum.meta_proxy.status import MetaProxyRoutingReceipt

if TYPE_CHECKING:
    from cognitum.agentic import Credential
    from cognitum.meta_proxy.config import MetaProxyClientConfig

_PRODUCT = "meta-proxy"
#: Required scope for the data-plane inference operations (distinct from
#: ``client.py``'s ``"meta-proxy.status"`` read scope).
_INFERENCE_SCOPE = "meta-proxy.inference"

#: Wire keys for the routing/upstream receipts embedded in a response body
#: (snake_case, consistent with the rest of the Proxy wire surface).
_ROUTING_RECEIPT_KEY = "cognitum_routing_receipt"
_UPSTREAM_RECEIPT_KEY = "cognitum_upstream_receipt"

#: The §D7 cloud-forwarding allowlist. Mapping is lowercase -> canonical
#: casing so a caller-supplied header of any case is normalized, and anything
#: NOT in this set (``Authorization``, ``Host``, ``Content-Length``, sponsor
#: markers, installation identity, training consent, ...) is dropped before
#: the request is built.
_FORWARD_ALLOWLIST: dict[str, str] = {
    h.lower(): h
    for h in (
        "Idempotency-Key",
        "X-Request-ID",
        "traceparent",
        "tracestate",
        "X-Cognitum-Fallback-Policy",
        "X-Cognitum-Min-Tier",
        "X-Cognitum-Max-Tier",
        "X-Cognitum-Escalation",
        "X-Cognitum-Cache",
        "X-Cognitum-Safety",
        "X-Cognitum-Sub-Tenant",
        "anthropic-version",
        "anthropic-beta",
    )
}


def filter_forwardable_headers(headers: Mapping[str, str] | None) -> dict[str, str]:
    """Return only the §D7-allowlisted subset of ``headers``, normalized to
    canonical casing. Everything else -- crucially ``Authorization`` and the
    local bearer -- is dropped, so a caller can never inject an auth override
    through the forwarding bag (ADR-0025a §D7).
    """
    result: dict[str, str] = {}
    if not headers:
        return result
    for name, value in headers.items():
        canonical = _FORWARD_ALLOWLIST.get(name.lower())
        if canonical is not None:
            result[canonical] = value
    return result


def parse_meta_proxy_routing_receipt(raw: Any) -> MetaProxyRoutingReceipt | None:
    """Decode a ``cognitum_routing_receipt`` body field into the typed
    receipt, or ``None`` when absent. Permissive about unknown fields, like
    the rest of the not-yet-contracted Proxy wire surface (§D4).
    """
    if not isinstance(raw, dict):
        return None
    return MetaProxyRoutingReceipt(
        request_id=str(raw.get("request_id", "")),
        configured_plane=str(raw.get("configured_plane", "")),
        selected_plane=str(raw.get("selected_plane", "")),
        automatic=bool(raw.get("automatic", False)),
        degraded=bool(raw.get("degraded", False)),
        routing_reason=raw.get("routing_reason"),
        workload_policy=raw.get("workload_policy"),
        consent_evidence_id=raw.get("consent_evidence_id"),
        upstream_receipt=raw.get("upstream_receipt"),
        local_usage=raw.get("local_usage"),
        warnings=raw.get("warnings"),
    )


async def _require_bearer(config: MetaProxyClientConfig, operation: str) -> Credential:
    from cognitum.agentic import CredentialRequest

    provider = config.local_credential_provider
    if provider is None:
        raise AgenticError(
            "authentication",
            f"MetaProxyClient.{operation} requires a local_credential_provider "
            "(ADR-0025a §D6: the bearer comes only from validated local state)",
            product=_PRODUCT,
            operation=operation,
            retryable=False,
        )
    return await provider.acquire(
        CredentialRequest(
            product=_PRODUCT,
            normalized_origin=config.origin,
            audience=config.origin,
            required_scopes=[_INFERENCE_SCOPE],
            operation=operation,
            interactive_allowed=False,
        )
    )


def _apply_bearer(headers: dict[str, str], credential: Credential) -> None:
    # ADR-0025a §D6/§D7: the local bearer is derived from validated local
    # state and placed by the SDK -- it is never a caller-forwarded header.
    if credential.scheme.lower() == "bearer":
        headers["Authorization"] = f"Bearer {credential.secret.reveal()}"
    else:
        headers[credential.scheme] = credential.secret.reveal()


async def _send_once(
    config: MetaProxyClientConfig,
    transport: httpx.AsyncClient,
    path: str,
    operation: str,
    body: dict[str, Any],
    credential: Credential,
    forwarded_headers: dict[str, str],
    idempotency_key: str,
) -> tuple[dict[str, Any], MetaProxyResponseMeta]:
    """One HTTP attempt. Never retries by itself -- the caller owns that."""
    request_id = str(uuid.uuid4())
    telemetry = config.telemetry
    if telemetry is not None:
        telemetry.on_request_start(operation, request_id)
    started_at = time.monotonic()

    # Start from the allowlisted forward headers, then overlay SDK-owned
    # headers so they always win (and so nothing caller-supplied can shadow
    # Accept/Content-Type/Authorization/request-id/idempotency-key).
    headers = dict(forwarded_headers)
    headers["Accept"] = "application/json"
    headers["Content-Type"] = "application/json"
    headers["X-Cognitum-Request-Id"] = request_id
    headers["Idempotency-Key"] = idempotency_key

    # ADR-0025a §D6/§D10: the bearer is sent ONLY to literal loopback (or an
    # explicitly opted-in non-loopback origin). Re-verify at the attachment
    # point rather than trusting construction alone.
    from cognitum.meta_proxy.config import bearer_target_allowed

    if not bearer_target_allowed(config.origin, config.allow_non_loopback):
        raise AgenticError(
            "authentication",
            f'{operation} refused to attach a local bearer to non-loopback origin '
            f'"{config.origin}" (ADR-0025a §D6/§D10: set allow_non_loopback only for '
            "the dangerous-preview remote case)",
            product=_PRODUCT,
            operation=operation,
            retryable=False,
        )
    _apply_bearer(headers, credential)

    url = f"{config.origin}{path}"
    try:
        response = await transport.post(url, headers=headers, json=body)
    except httpx.HTTPError as cause:
        raise AgenticError(
            "transport",
            f"{operation} request failed: {cause}",
            product=_PRODUCT,
            operation=operation,
            request_id=request_id,
            retryable=True,
            cause=cause,
        ) from cause

    duration_ms = (time.monotonic() - started_at) * 1000
    retry_after_header = response.headers.get("retry-after")
    retry_after_ms = int(float(retry_after_header) * 1000) if retry_after_header else None

    if telemetry is not None:
        from cognitum.meta_proxy.config import MetaProxyTelemetryEvent

        telemetry.on_request_end(
            MetaProxyTelemetryEvent(
                operation=operation,
                request_id=request_id,
                http_status=response.status_code,
                duration_ms=duration_ms,
                retry_after_ms=retry_after_ms,
            )
        )

    # ADR-0025a §D10: redirects are rejected outright, never followed. With
    # ``follow_redirects=False`` httpx returns the 3xx as-is, so we must turn
    # it into a hard protocol error rather than silently treating the (empty)
    # redirect body as a success.
    if 300 <= response.status_code < 400:
        location = response.headers.get("location")
        raise AgenticError(
            "protocol",
            f"{operation} received a {response.status_code} redirect"
            + (f' to "{location}"' if location else "")
            + " -- the Meta Proxy transport rejects redirects (ADR-0025a §D10)",
            product=_PRODUCT,
            operation=operation,
            status=response.status_code,
            request_id=request_id,
            retryable=False,
        )

    if response.status_code >= 400:
        err = map_meta_proxy_http_error(response, operation, request_id)
        if err.retry_after_ms is None:
            err.retry_after_ms = retry_after_ms
        raise err

    data: dict[str, Any] = response.json()
    routing_receipt = parse_meta_proxy_routing_receipt(data.get(_ROUTING_RECEIPT_KEY))
    upstream_receipt = data.get(_UPSTREAM_RECEIPT_KEY)
    meta = MetaProxyResponseMeta(
        request_id=response.headers.get("x-cognitum-request-id", request_id),
        http_status=response.status_code,
        product_version=response.headers.get("x-cognitum-product-version"),
        protocol_version=response.headers.get("x-cognitum-protocol-version"),
        retry_after=float(retry_after_header) if retry_after_header else None,
        routing_receipt=routing_receipt,
        upstream_receipt=upstream_receipt,
        unknown_headers=collect_unknown_headers(response.headers),
    )
    return data, meta


async def post_chat_forwarding(
    config: MetaProxyClientConfig,
    transport: httpx.AsyncClient,
    path: str,
    operation: str,
    body: dict[str, Any],
    caller_headers: Mapping[str, str] | None,
) -> tuple[dict[str, Any], MetaProxyResponseMeta]:
    """Forward one non-streaming inference POST through the Proxy (§D7).

    Attaches ONLY the §D7-allowlisted subset of ``caller_headers`` plus the
    SDK-owned headers and the validated local bearer; generates an
    idempotency key when the caller did not supply one; performs at most one
    401 credential refresh; rejects redirects; and decodes the routing/
    upstream receipts into the returned metadata.

    Per §D8, a Proxy POST is never automatically retried on 429/502/503 (or
    any other status): the currently-deployed Proxy drops the caller's
    ``Idempotency-Key`` server-side, so attaching one client-side does not
    make a silent retry safe against duplicate spend. Any such error
    surfaces as a single terminal, non-retryable ``AgenticError`` whose
    ``retry_after_ms`` (populated from a ``Retry-After`` response header,
    when present) lets the CALLER decide whether to retry manually.
    """
    forwarded = filter_forwardable_headers(caller_headers)
    # Reuse a caller-supplied (allowlisted) Idempotency-Key when present so
    # the retry replay is safe against the same logical request; otherwise
    # generate one. Stable across every retry of this call.
    idempotency_key = forwarded.pop("Idempotency-Key", None) or str(uuid.uuid4())

    credential = await _require_bearer(config, operation)

    refreshed_once = False

    while True:
        try:
            return await _send_once(
                config,
                transport,
                path,
                operation,
                body,
                credential,
                forwarded,
                idempotency_key,
            )
        except AgenticError as err:
            if err.status == 401 and not refreshed_once:
                refreshed_once = True
                if config.local_credential_provider is not None:
                    await config.local_credential_provider.invalidate(
                        "401 challenge from meta-proxy"
                    )
                credential = await _require_bearer(config, operation)
                continue

            # ADR-0025a §D8: "No Proxy POST is automatically retried while it
            # drops `Idempotency-Key`" -- the currently-deployed Proxy does not
            # honor the header for server-side dedup, so attaching one here
            # does not make a retry safe. 429/502/503 (and every other status)
            # surface as a single terminal error; `err.retry_after_ms` lets the
            # caller retry manually.
            raise


__all__ = [
    "filter_forwardable_headers",
    "parse_meta_proxy_routing_receipt",
    "post_chat_forwarding",
]

File: sdks/rust/src/meta_llm/nonstream.rs
//! Non-streaming HTTP call logic for the two "direct nonstream call[s]
//! whose accepted contract declares safe replay" named in ADR-0024a §D7:
//! `chat.completions` and `messages.create`. Issue #58 / M2 continuation.
//!
//! Deliberately out of scope here: streaming/SSE parsing (§D5) -- folded
//! into `client.rs`/this module in later passes.
//!
//! ADR-0024b D11 migration step 1 (issue #59): this module now also
//! decodes a `MetaLlmReceipt` from the response body when the server
//! includes one (`cognitum_receipt`). `routing_controls` validation
//! happens in `client.rs` before `serde_json::to_value` even runs, not
//! here -- this module only sees the already-serialized `Value` body.
//!
//! Retry (ADR-0023 §D3/§D4, ADR-0024a §D6):
//! - 401: at most one credential refresh after a verified challenge (the
//!   401 response itself), then retry once with the same idempotency key
//!   and body. A second 401 is returned as-is.
//! - 429/502/503: bounded retry using the frozen `RetryPolicy` /
//!   `equal_jitter_delay_ms` (ADR-0005/ADR-0023 verbatim), gated on the
//!   idempotency-with-key binding built in `./idempotency.rs` — this is
//!   what makes the replay safe.
//! - 400/403/404/409/402/422 and anything else: never retried.

use std::time::{Duration, SystemTime, UNIX_EPOCH};

use serde_json::Value;
use uuid::Uuid;

use crate::agentic::{
    assert_scope_granted, equal_jitter_delay_ms, AgenticError, AgenticErrorKind, Credential,
    CredentialRequest, RetryPolicy,
};

use super::client::MetaLlmClient;
use super::config::MetaLlmTelemetryEvent;
use super::envelope::MetaLlmResponseMeta;
use super::idempotency::{build_idempotency_binding, canonical_request_sha256};
use super::types::parse_meta_llm_receipt;
use super::PRODUCT;

/// Required scope for the inference-serving operations in this pass.
/// Distinct from `get_json`'s `"meta-llm.read"` — these are mutating
/// generation calls, not discovery reads (ADR-0024a §D8).
const INFERENCE_SCOPE: &str = "meta-llm.inference";

/// Per-operation required-scope map for ADR-0022 §D5's scope preflight,
/// covering every "completion-family route" per ADR-0024a §D8 that shares
/// this module's `post_json_idempotent`/credential path.
///
/// PROVISIONAL: no ADR-0020 OpenAPI/JSON-Schema contract bundle publishing
/// a real scope-token vocabulary exists yet (ADR-0024a §D9 gate #1/#8), so
/// every completion-family operation maps to the same literal
/// `INFERENCE_SCOPE` already used in the `CredentialRequest` sent to
/// `acquire()` below — this names the mapping explicitly so a real
/// per-operation vocabulary can slot in later without changing the
/// preflight call site.
fn required_scope_for(operation: &str) -> &'static str {
    match operation {
        "chat.completions" | "chat_completions_stream" | "messages.create"
        | "messages.count_tokens" | "completions" | "responses" | "embeddings" => {
            INFERENCE_SCOPE
        }
        _ => INFERENCE_SCOPE,
    }
}

/// Cheap non-cryptographic jitter in `[0, bound]`, matching the existing
/// cloud `Client`'s `retry_hint::pseudo_jitter_ms` convention (subsecond
/// `SystemTime` entropy) rather than pulling in a `rand` dependency for a
/// backoff jitter that only needs to avoid lockstep retries across
/// callers, not cryptographic unpredictability.
/// `pub(super)` so `./stream/chat_completions_stream.rs` can reuse this
/// verbatim (issue #58 D5 streaming pass).
pub(super) fn random_jitter_ms(bound: u64) -> u64 {
    if bound == 0 {
        return 0;
    }
    let nanos = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.subsec_nanos() as u64)
        .unwrap_or(0);
    nanos % (bound + 1)
}

impl MetaLlmClient {
    /// Shared idempotent-with-key POST used by `chat_completions` and
    /// `messages_create`. Returns the raw JSON body plus response
    /// metadata; callers deserialize into their own protocol-specific
    /// response type.
    #[allow(clippy::result_large_err)]
    pub(super) async fn post_json_idempotent(
        &self,
        path: &str,
        operation: &str,
        body: Value,
    ) -> Result<(Value, MetaLlmResponseMeta), AgenticError> {
        let mut credential = self.require_credential(operation).await?;
        let idempotency_key = Uuid::new_v4().to_string();
        let canonical_sha256 = canonical_request_sha256(&body);
        let tenant = self
            .config
            .default_request_context
            .as_ref()
            .and_then(|rc| rc.tenant.as_ref());

        let retry_policy = RetryPolicy::default();
        let mut attempt: u32 = 0;
        let mut sleep_budget_used_ms: u64 = 0;
        let mut refreshed_once = false;

        loop {
            let binding = build_idempotency_binding(
                operation,
                path,
                &credential,
                tenant,
                canonical_sha256.clone(),
                idempotency_key.clone(),
            );

            match self
                .send_post_once(path, operation, &body, &credential, &binding.idempotency_key)
                .await
            {
                Ok(ok) => return Ok(ok),
                Err(err) => {
                    if err.status == Some(401) && !refreshed_once {
                        refreshed_once = true;
                        if let Some(provider) = self.config.credential_provider.as_ref() {
                            provider.invalidate("401 challenge from meta-llm").await;
                        }
                        credential = self.require_credential(operation).await?;
                        continue;
                    }

                    let is_bounded_retryable =
                        matches!(err.status, Some(429) | Some(502) | Some(503));
                    if is_bounded_retryable && attempt + 1 < retry_policy.max_attempts {
                        let server_hint_ms = err.retry_after_ms.unwrap_or(0);
                        let jitter_ms = random_jitter_ms(retry_policy.base_ms);
                        let delay_ms = equal_jitter_delay_ms(
                            attempt,
                            &retry_policy,
                            server_hint_ms,
                            jitter_ms,
                        );
                        if sleep_budget_used_ms.saturating_add(delay_ms)
                            > retry_policy.retry_sleep_budget_ms
                        {
                            return Err(err);
                        }
                        sleep_budget_used_ms += delay_ms;
                        tokio::time::sleep(Duration::from_millis(delay_ms)).await;
                        attempt += 1;
                        continue;
                    }

                    return Err(err);
                }
            }
        }
    }

    /// Resolve a credential, failing closed with `Authentication` when no
    /// provider is configured — mirrors `http.rs`'s `get_json` gate for
    /// `whoami`/`models`, but with the inference scope (ADR-0024a §D8).
    /// `pub(super)` (rather than private) so `./stream/chat_completions_stream.rs`
    /// can reuse it verbatim (issue #58 D5 streaming pass) instead of
    /// duplicating credential-acquisition logic.
    #[allow(clippy::result_large_err)]
    pub(super) async fn require_credential(&self, operation: &str) -> Result<Credential, AgenticError> {
        let Some(provider) = self.config.credential_provider.as_ref() else {
            return Err(AgenticError {
                product: Some(PRODUCT.to_owned()),
                operation: Some(operation.to_owned()),
                ..AgenticError::new(
                    AgenticErrorKind::Authentication,
                    format!("MetaLlmClient::{operation} requires a credential_provider"),
                )
            });
        };
        let required_scope = required_scope_for(operation);
        let request = CredentialRequest {
            product: PRODUCT.to_owned(),
            normalized_origin: self.config.base_url.clone(),
            audience: self.config.base_url.clone(),
            required_scopes: vec![required_scope.to_owned()],
            operation: operation.to_owned(),
            interactive_allowed: false,
        };
        let credential = provider.acquire(&request).await.map_err(|mut e| {
            if e.product.is_none() {
                e.product = Some(PRODUCT.to_owned());
            }
            e
        })?;
        // ADR-0022 §D5 scope preflight: fail closed BEFORE any I/O when the
        // credential's granted scopes are known and insufficient. A
        // credential with unknown (`None`) granted scopes — e.g.
        // `StaticApiKeyCredentialProvider`'s today — is sent through
        // unchecked; the server remains authoritative for that case.
        assert_scope_granted(PRODUCT, operation, required_scope, &credential)?;
        Ok(credential)
    }

    /// One HTTP attempt. Never retries by itself — the caller
    /// (`post_json_idempotent`) owns retry/backoff decisions.
    #[allow(clippy::result_large_err)]
    async fn send_post_once(
        &self,
        path: &str,
        operation: &str,
        body: &Value,
        credential: &Credential,
        idempotency_key: &str,
    ) -> Result<(Value, MetaLlmResponseMeta), AgenticError> {
        let request_id = Uuid::new_v4().to_string();
        if let Some(telemetry) = self.config.telemetry.as_ref() {
            telemetry.on_request_start(operation, &request_id);
        }
        let started_at = std::time::Instant::now();

        let mut headers = reqwest::header::HeaderMap::new();
        headers.insert(
            reqwest::header::ACCEPT,
            reqwest::header::HeaderValue::from_static("application/json"),
        );
        headers.insert(
            reqwest::header::CONTENT_TYPE,
            reqwest::header::HeaderValue::from_static("application/json"),
        );
        if let Ok(value) = reqwest::header::HeaderValue::from_str(&request_id) {
            headers.insert("X-Cognitum-Request-Id", value);
        }
        // ADR-0024a §D7 / ADR-0005: the caller-attested idempotency key,
        // stable across every retry of one logical call.
        if let Ok(value) = reqwest::header::HeaderValue::from_str(idempotency_key) {
            headers.insert("Idempotency-Key", value);
        }
        self.apply_auth(&mut headers, credential);

        let url = format!("{}{}", self.config.base_url, path);
        let response = self
            .http
            .post(&url)
            .headers(headers)
            .json(body)
            .send()
            .await
            .map_err(|cause| AgenticError {
                product: Some(PRODUCT.to_owned()),
                operation: Some(operation.to_owned()),
                request_id: Some(request_id.clone()),
                retryable: true,
                ..AgenticError::new(
                    AgenticErrorKind::Transport,
                    format!("{operation} request failed: {cause}"),
                )
            })?;

        let status = response.status();
        let duration_ms = started_at.elapsed().as_millis() as u64;

        let protocol_version = response
            .headers()
            .get("x-cognitum-protocol-version")
            .and_then(|v| v.to_str().ok())
            .map(str::to_owned);
        let response_request_id = response
            .headers()
            .get("x-cognitum-request-id")
            .and_then(|v| v.to_str().ok())
            .map(str::to_owned)
            .unwrap_or_else(|| request_id.clone());
        let idempotent_replay = response
            .headers()
            .get("x-cognitum-idempotent-replay")
            .and_then(|v| v.to_str().ok())
            .map(|v| v.eq_ignore_ascii_case("true"));
        let retry_after_header_ms = response
            .headers()
            .get(reqwest::header::RETRY_AFTER)
            .and_then(|v| v.to_str().ok())
            .and_then(|v| v.parse::<f64>().ok())
            .map(|secs| (secs * 1000.0) as u64);

        if let Some(telemetry) = self.config.telemetry.as_ref() {
            telemetry.on_request_end(&MetaLlmTelemetryEvent {
                operation: operation.to_owned(),
                request_id: request_id.clone(),
                http_status: Some(status.as_u16()),
                duration_ms: Some(duration_ms),
                retry_after_ms: retry_after_header_ms,
                idempotent_replay,
            });
        }

        if !status.is_success() {
            let body_text = response.text().await.unwrap_or_default();
            let mut err =
                MetaLlmClient::map_http_error(status, &body_text, operation, &response_request_id);
            if err.retry_after_ms.is_none() {
                err.retry_after_ms = retry_after_header_ms;
            }
            return Err(err);
        }

        let data: Value = response.json().await.map_err(|cause| AgenticError {
            product: Some(PRODUCT.to_owned()),
            operation: Some(operation.to_owned()),
            request_id: Some(response_request_id.clone()),
            ..AgenticError::new(
                AgenticErrorKind::Protocol,
                format!("{operation} response body was not valid JSON: {cause}"),
            )
        })?;

        // ADR-0024b §D3/§D11 step 1: decode a `cognitum_receipt` field
        // embedded in the response body, if present, into the typed
        // `MetaLlmReceipt` -- same wire key the SSE path already
        // recognizes (`stream/openai_events.rs`).
        let receipt = data
            .as_object()
            .and_then(|obj| obj.get("cognitum_receipt"))
            .and_then(parse_meta_llm_receipt);

        let meta = MetaLlmResponseMeta {
            request_id: response_request_id,
            http_status: status.as_u16(),
            protocol_version,
            retry_after_ms: retry_after_header_ms,
            idempotent_replay,
            receipt,
            warnings: None,
            unknown_headers: None,
        };
        Ok((data, meta))
    }
}
